### PR TITLE
op-batcher: wait for node sync & check recent L1 txs at startup to avoid duplicate txs

### PR DESCRIPTION
**Pull request by bitwiseguy**
**Description**

During startup, `op-batcher` should check to see if it recently submitted any batcher txs to L1. If it has submitted some that are on L1 but not yet finalized, it should wait for them to finalize instead of sending a duplicate batcher tx to post the same information to L1. This functionality is initially being implemented as a feature flag for the batcher: `CheckRecentTxsDepth`.

**Tests**

New test cases added. Defined [here](https://github.com/ethereum-optimism/optimism/pull/10193/files#diff-c10f0c550c104748ac70c7ee8da29955eded53dd27d218b004bd8e62fe05191c).

**Additional context**

Implementation steps:

1. Use `op-batcher` new cli flag `CheckRecentTxsDepth` to determine if we should look for recent batcher txs during startup. If set, the value determines how many blocks to look back in time.

2. Check to see if the batcher's nonce has changed in the last `CheckRecentTxsDepth` blocks. If a nonce diff is detected, wait until the latest batcher tx is finalized (i.e. the batcher syncs up to that block). If no nonce diff detected or the `CheckRecentTxsDepth` is set to 0, the batcher will sync up to the current l1 block tip during startup.

**Metadata**

- Fixes https://github.com/ethereum-optimism/client-pod/issues/697
- Fixes https://github.com/ethereum-optimism/client-pod/issues/696
- Fixes https://github.com/ethereum-optimism/client-pod/issues/698

### CODE REVIEW

File: op-batcher/batcher/driver.go [hunk 1]
```diff
@@ -324,6 +329,71 @@ func (l *BatchSubmitter) loop() {
 	}
 }
 
+// checkRecentTxsOnStart Check to see if there was a batcher tx sent recently that
+// still needs more block confirmations before being considered finalized
+func (l *BatchSubmitter) checkRecentTxsOnStart() {
+	if !l.Config.CheckRecentTxsOnStart {
```
Comment (tynes) [inline op-batcher/batcher/driver.go]: Do you think its better to have this outside of the function rather than inside? Feels like its easier to test that way

File: op-batcher/batcher/driver.go [hunk 2]
```diff
@@ -324,6 +329,71 @@ func (l *BatchSubmitter) loop() {
...
+		return
+	}
+
+	currentBlock, err := l.Txmgr.BlockNumber(l.shutdownCtx)
+	if err != nil {
+		l.Log.Error("failed to retrieve current block number", "err", err)
+		return
+	}
+
+	currentBlockBig := new(big.Int)
```
Comment (tynes) [inline op-batcher/batcher/driver.go]: personal style nit: why add the type to the variable name? just seems like unnecessary overhead. also can set the uint64 on the same line to reduce a loc

**Comment by tynes** [comment]
It seems quite unreliable to use the mempool as there is no concept of a global mempool? The mempool probably will go away in the future anyways, its just a partial view on the set of transactions that can be included. A more robust solution would be to have a db of sent txs and use that as a source of truth but that removes the nice stateless property of the batcher

edit: i now see this isn't built with the mempool

File: op-batcher/batcher/driver.go [hunk 3]
```diff
@@ -324,6 +329,71 @@ func (l *BatchSubmitter) loop() {
 	}
 }
 
+// checkRecentTxsOnStart Check to see if there was a batcher tx sent recently that
+// still needs more block confirmations before being considered finalized
+func (l *BatchSubmitter) checkRecentTxsOnStart() {
```
Comment (tynes) [inline op-batcher/batcher/driver.go]: This requires archival node support

Comment (tynes) [reply op-batcher/batcher/driver.go]: We need to be mindful about requiring archival node support, i don't believe it was required previously but we shouldn't add the requirement unless its communicated that its now being added and everybody knows and its no issue to not use an archive node

Comment (bitwiseguy) [reply op-batcher/batcher/driver.go]: Great point. With this function we are only attempting to look back in time as far as the `currentBlockNum - confirmationDepth` which _should_ only be ~5 blocks behind the head of the L1 chain. If `confirmationDepth` is only set to 5, would that still require archival node support? Do regular nodes have any guarantees about how far back in time they will service `NonceAt` requests or must we assume they won't service those requests at all?

File: op-batcher/batcher/driver.go (hunk 1 above)
Comment (bitwiseguy) [reply op-batcher/batcher/driver.go]: Yes that would make it easier to test, at least at the unit test level that this PR implements. Will make that change

File: op-batcher/batcher/driver.go (hunk 3 above)
Comment (trianglesphere) [reply op-batcher/batcher/driver.go]: Normal nodes usually have ~128 blocks of full state available

Comment (ajsutton) [reply op-batcher/batcher/driver.go]: Would it be reasonable and simpler to just wait for the op-node instance sync to reach the L1 head the batcher observes at startup? Then any already included transactions would be reflected in its state.  Checking the nonce can shortcircuit this if there are no recent transactions but with a typical follow distance of 5 blocks it seems reasonable to effectively wait 1 minute after startup to ensure the node is in sync.

Also my intuition (maybe I misunderstood) is that it's the follow distance of op-node that matters here, not the confirmation depth of the batcher. The batcher starts from the op-node `safe` head and submits blocks up to its `unsafe` head, so as long as the op-node has processed the batcher's previous transaction, the batcher won't try to resubmit those blocks.

File: op-batcher/batcher/driver.go (hunk 2 above)
Comment (bitwiseguy) [reply op-batcher/batcher/driver.go]: Updated in https://github.com/ethereum-optimism/optimism/pull/10193/commits/add2881fd459ad8c34fb94ffe0b70802a6f2b1b1

File: op-batcher/batcher/driver.go (hunk 1 above)
Comment (bitwiseguy) [reply op-batcher/batcher/driver.go]: Updated in https://github.com/ethereum-optimism/optimism/pull/10193/commits/add2881fd459ad8c34fb94ffe0b70802a6f2b1b1

File: op-batcher/batcher/config.go [hunk 4]
```diff
@@ -67,7 +67,8 @@ type CLIConfig struct {
 	// Type of compressor to use. Must be one of [compressor.KindKeys].
 	Compressor string
 
-	Stopped bool
+	Stopped        bool
+	CheckRecentTxs bool
```
Comment (sebastianst) [inline op-batcher/batcher/config.go]: Would be good to follow the convention in this CLI struct to describe briefly what this does (and take the opportunity to also do this for `Stopped`).

I also suggest we add another feature flag like `WaitNodeSync` to enable the node sync waiting with `WaitRollupSync` at all.

```suggestion
    // If Stopped is true, the batcher starts stopped and won't start batching right away.
    // Batching needs to be started via an admin RPC.
	Stopped bool
	
	// Whether to wait for the sequencer to sync to a recent block at startup.
	WaitNodeSync bool
	
	// Whether to check recent batcher transactions when waiting for node sync at startup.
	// If false, the batcher will just use the current head.
	CheckRecentTxs bool
```

File: op-batcher/batcher/driver.go [hunk 5]
```diff
@@ -283,6 +285,11 @@ func (l *BatchSubmitter) loop() {
 			l.Log.Info("Txmgr is closed, remaining channel data won't be sent")
 		}
 	}
+
+	if !l.Config.CheckRecentTxsOnStart {
+		l.checkRecentTxsOnStart()
+	}
```
Comment (sebastianst) [inline op-batcher/batcher/driver.go]: I suggest we move the startup check to the top of `loop` right after `defer l.wg.Done()` and return on an error, because if it fails, something is wrong with the node we're using and it's ok to fail the program.

We should then also let `checkRecentTxsOnStart`/`waitNodeSync` return an `error`, so we error log it here and return the loop.

I also suggest in another comment that we feature-flag the node sync waiting separately, and in it branch on the `CheckRecentTxs` config, so this outer check would become
```go
	if !l.Config.WaitNodeSync{
		l.waitNodeSync()
	}
```

File: op-batcher/batcher/driver.go (hunk 3 above)
Comment (sebastianst) [reply op-batcher/batcher/driver.go]: @tynes Yes, the idea is to only look back the op-node's confirmation depth, so ~5 blocks, which any EL node should be able to handle, as @trianglesphere said.

@ajsutton yes that's correct, the node's conf depth ("follow distance") matters. The function `WaitRollupSync` that is invoked at the end takes care of this, it waits for the rollup node to sync its `CurrentL1` to the provided block number (I make a separate review comment there because we don't need to add the batcher's conf depth). The `CurrentL1` is the latest L1 block from which it derived safe blocks from, so exactly what we need.

Using the recent head from when the batcher starts up and pass it to `WaitRollupSync` was meant to be the first iteration (https://github.com/ethereum-optimism/client-pod/issues/698), and this implementation by Sam (https://github.com/ethereum-optimism/client-pod/issues/697) to be an optimization as to not unnecessarily wait for a minute at startup if there weren't any recent batcher txs. Note that if only the batcher (but not op-node) restarts, the node would typically be in sync, so it would start its work right away, which might be quite helpful in incident scenarios where the batcher is restarted and we want it to quickly start batching.

I also propose to feature-flag this behavior separately.

File: op-batcher/batcher/service.go [hunk 6]
```diff
@@ -72,7 +73,8 @@ type BatcherService struct {
 	balanceMetricer io.Closer
 	stopped         atomic.Bool
 
-	NotSubmittingOnStart bool
+	NotSubmittingOnStart  bool
+	CheckRecentTxsOnStart bool
```
Comment (sebastianst) [inline op-batcher/batcher/service.go]: We already embed the `BatcherConfig`, no need to add this field.

File: op-batcher/batcher/service.go [hunk 7]
```diff
@@ -41,7 +41,8 @@ type BatcherConfig struct {
 
 	// UsePlasma is true if the rollup config has a DA challenge address so the batcher
 	// will post inputs to the Plasma DA server and post commitments to blobs or calldata.
-	UsePlasma bool
+	UsePlasma             bool
+	CheckRecentTxsOnStart bool
```
Comment (sebastianst) [inline op-batcher/batcher/service.go]: See other comment at CLI config, we should also add a field `WaitNodeSync bool`. I also suggest we keep the field CLI config names, so
```suggestion
	
	CheckRecentTxs bool
	WaitNodeSync   bool
```

It would also be more readable to not group them with `UsePlasma`, but instead either in their own group, or above with the others networking configuration fields.

File: op-batcher/batcher/service.go [hunk 8]
```diff
@@ -96,6 +98,7 @@ func (bs *BatcherService) initFromCLIConfig(ctx context.Context, version string,
 	bs.PollInterval = cfg.PollInterval
 	bs.MaxPendingTransactions = cfg.MaxPendingTransactions
 	bs.NetworkTimeout = cfg.TxMgrConfig.NetworkTimeout
+	bs.CheckRecentTxsOnStart = cfg.CheckRecentTxs
```
Comment (sebastianst) [inline op-batcher/batcher/service.go]: would set the 2nd field here.

File: op-batcher/flags/flags.go [hunk 9]
```diff
@@ -127,6 +127,13 @@ var (
 		Value:   2 * time.Minute,
 		EnvVars: prefixEnvVars("ACTIVE_SEQUENCER_CHECK_DURATION"),
 	}
+	CheckRecentTxsFlag = &cli.BoolFlag{
```
Comment (sebastianst) [inline op-batcher/flags/flags.go]: would add 2nd flag `WaitNodeSyncFlag` here.

File: op-node/rollup/types.go [hunk 10]
```diff
@@ -122,6 +123,8 @@ type Config struct {
 
 	// UsePlasma is activated when the chain is in plasma mode.
 	UsePlasma bool `json:"use_plasma"`
+
+	VerifierConfDepth uint64 `json:"verifier_conf_depth"`
```
Comment (sebastianst) [inline op-node/rollup/types.go]: I see that we need this to know how far to look back... it's not exactly in the same category as the other fields, which are more like protocol parameter, whereas this is a node runtime config parameter. @trianglesphere wdyt about adding this field here?

File: op-batcher/batcher/driver.go [hunk 11]
```diff
@@ -324,6 +331,66 @@ func (l *BatchSubmitter) loop() {
...
+		currentBlock.Sub(currentBlock, big.NewInt(1))
+		currentNonce, err = l.L1Client.NonceAt(l.shutdownCtx, l.Txmgr.From(), currentBlock)
+		if err != nil {
+			l.Log.Error("Failed to retrieve nonce", "err", err)
+			return
+		}
+	}
+
+	batcherTxFinalizedBlock := currentBlock.Uint64() + 1 + blockConfirms
+	err = dial.WaitRollupSync(l.shutdownCtx, l.Log, rollupClient, batcherTxFinalizedBlock, time.Second*5)
```
Comment (sebastianst) [inline op-batcher/batcher/driver.go]:
```suggestion
	batcherTxBlock := currentBlock.Uint64() + 1
	err = dial.WaitRollupSync(l.shutdownCtx, l.Log, rollupClient, batcherTxBlock, time.Second*12)
```
We don't need to handle the verifier confirmation depth, because `WaitRollupSync` waits for the `CurrentL1` block to catch up, which already is `VerifierConfDepth` deep.

Since we're waiting on L1 blocks, it makes sense to poll every 12 sec.

File: op-batcher/batcher/driver.go [hunk 12]
```diff
@@ -324,6 +331,66 @@ func (l *BatchSubmitter) loop() {
 	}
 }
 
+// checkRecentTxsOnStart Check to see if there was a batcher tx sent recently that
+// still needs more block confirmations before being considered finalized
+func (l *BatchSubmitter) checkRecentTxsOnStart() {
```
Comment (sebastianst) [inline op-batcher/batcher/driver.go]: See other comment, I suggest we return an error here instead, and then in the function don't error-log, but return errors.

I also suggest we rename this function like
`func (l *BatchSubmitter) waitNodeSync() error`
and in it either set `batcherTxBlock ` to `l.l1Tip()` if `CheckRecentTxs` is false, or do the actual recent tx checking if it's true.

I also suggest we move the whole recent tx checking code into its own function (similar to `WaitRollupSync`) so we can potentially reuse it, like
```go
// CheckRecentTxs checks the depth recent blocks for transactions form the account with address addr
// and returns the most recent block and true, if any was found, or the oldest block checked and false, if not. 
func CheckRecentTxs(
	ctx context.Context,
	lgr log.Logger,
	l1 L1Client,
	depth int,
	addr common.Address,
) (recentBlock uint64, found bool, err error)
```
We could move this function into package `op-service/eth`.

File: op-batcher/batcher/driver.go [hunk 13]
```diff
@@ -324,6 +331,66 @@ func (l *BatchSubmitter) loop() {
...
+	currBlock, err := l.Txmgr.BlockNumber(l.shutdownCtx)
```
Comment (sebastianst) [inline op-batcher/batcher/driver.go]: You can use `L1Client.HeaderByNumber(ctx, nil)`, which removes the dependency on the `TxMgr`. Note that I suggest in another comment to move this logic into its own reusable function `CheckRecentTxs` in package `op-service/eth`. If we were to keep it in the batcher, you could also just use `BatchSubmitter.l1Tip`.

File: op-batcher/batcher/driver.go [hunk 14]
```diff
@@ -324,6 +331,66 @@ func (l *BatchSubmitter) loop() {
...
+	}
+	rollupConfig, err := rollupClient.RollupConfig(l.shutdownCtx)
+	if err != nil {
+		l.Log.Error("Failed to retrieve rollup config", "err", err)
+		return
+	}
+	blockConfirms := rollupConfig.VerifierConfDepth
+
+	previousBlockBig := new(big.Int)
+	previousBlockBig.Sub(currentBlock, big.NewInt(int64(blockConfirms)))
```
Comment (sebastianst) [inline op-batcher/batcher/driver.go]:
```suggestion
	previousBlockBig := new(big.Int).Sub(currentBlock, big.NewInt(int64(blockConfirms)))
```
maybe we also call it something like `oldestBlock` because it's not the _previous_ block.

File: op-batcher/batcher/driver.go [hunk 15]
```diff
@@ -324,6 +331,66 @@ func (l *BatchSubmitter) loop() {
...
+	previousNonce, err := l.L1Client.NonceAt(l.shutdownCtx, l.Txmgr.From(), previousBlockBig)
+	if err != nil {
+		l.Log.Error("Failed to retrieve previous nonce", "err", err)
+		return
+	}
+
+	if currentNonce == previousNonce {
+		l.Log.Info("No recent batcher txs detected. No need to wait for rollup sync")
+		return
+	}
```
Comment (sebastianst) [inline op-batcher/batcher/driver.go]: we would still need to wait for node-sync on the oldest block, so not return here, but set the `batcherTxBlock` to the `oldestBlock`. Note that if we move this into its own function `CheckRecentTxs`, we `return oldestBlock, false, nil` here.

File: op-batcher/batcher/driver.go [hunk 16]
```diff
@@ -324,6 +331,66 @@ func (l *BatchSubmitter) loop() {
...
+	for currentNonce != previousNonce {
+		currentBlock.Sub(currentBlock, big.NewInt(1))
+		currentNonce, err = l.L1Client.NonceAt(l.shutdownCtx, l.Txmgr.From(), currentBlock)
+		if err != nil {
+			l.Log.Error("Failed to retrieve nonce", "err", err)
+			return
+		}
+	}
+
+	batcherTxFinalizedBlock := currentBlock.Uint64() + 1 + blockConfirms
```
Comment (sebastianst) [inline op-batcher/batcher/driver.go]: If moving to `CheckRecentTxs`, here we'd return `return currentBlock, true, nil`.

File: op-batcher/batcher/driver_test.go [hunk 17]
```diff
@@ -117,3 +142,92 @@ func TestBatchSubmitter_SafeL1Origin_FailsToResolveRollupClient(t *testing.T) {
 	_, err := bs.safeL1Origin(context.Background())
 	require.Error(t, err)
 }
+
+func TestBatchSubmitter_CheckRecentTxsOnStart(t *testing.T) {
+	bs, ep := setup(t)
+	txMgr := bs.Txmgr.(*mocks.TxManager)
+	l1Client := bs.L1Client.(*MockL1Client)
```
Comment (sebastianst) [inline op-batcher/batcher/driver_test.go]: Maybe we move these into `t.Run` to set up the batch submitter and mocks for individual tests? Then we don't need to reset mock fields and are guaranteed to have fresh mocks and a fresh BS. Note that if we move the recent trx logic into its own function `CheckRecentTxs`, the setup becomes a bit lighter.

**Review by sebastianst — CHANGES_REQUESTED** [review]

File: op-node/rollup/types.go (hunk 10 above)
Comment (sebastianst) [reply op-node/rollup/types.go]: An alternative could be to change the `CheckRecentTxs` bool config param to an int `CheckRecentTxsDepth` and specify the depth manually, with a default of `4` which is the current default verifier conf depth.

File: op-batcher/batcher/driver.go (hunk 13 above)
Comment (bitwiseguy) [reply op-batcher/batcher/driver.go]: I am currently also using `l.Txmgr.From()` to retrieve the batcher's eth address. Is there already a different way to get that address or do I need to implement one to completely remove the dependence on `l.Txmgr`?

edit: I see that you are saying to remove `checkRecentTxsOnStart` dependence on `l.Txmgr`, but the `l.waitNodeSync` function can still use `l.Txmgr.From()` to retrieve the address and pass it to `CheckRecentTxs` once `CheckRecentTxs` is moved into the `op-service/eth` package

File: op-node/rollup/types.go (hunk 10 above)
Comment (trianglesphere) [reply op-node/rollup/types.go]: Nice catch, I would like to avoid adding this parameter here. This config is what defines the L2 chain for the op-node & `verifier_conf_depth` is not a consensus parameter. We may want to expose this value a different way (maybe an extended config which is returned by the op-node RPC server).

File: op-batcher/batcher/driver.go (hunk 13 above)
Comment (sebastianst) [reply op-batcher/batcher/driver.go]: To your edit: yes that's correct.

I think for now it's easiest to get the batcher sender address with `l.Txmgr.From()` in `l.waitNodeSync` 👍🏻 I'm also not aware of a different way right now. The batcher configures the signer address, which is directly used to set up the txmgr.

File: op-node/rollup/types.go (hunk 10 above)
Comment (sebastianst) [reply op-node/rollup/types.go]: Thanks for confirming. I suggest in this iteration to just make the depth a config parameter, with default 0 to disable the recent tx check by default. If demand comes up to read this directly from op-node, we can add an rpc to op-node to retrieve some runtime config params, separate from the rollup config.

File: op-batcher/batcher/driver.go (hunk 16 above)
Comment (bitwiseguy) [reply op-batcher/batcher/driver.go]: Should `CheckRecentTxs` return the last block that contains a lower nonce? Or the first block that contains the current nonce? I am implementing the function calls like this and trying to avoid an "off-by-one bug":
```
l1TargetBlock, found, err := eth.CheckRecentTxs(l.shutdownCtx, l.Log, l.L1Client, int(rollupConfig.VerifierConfDepth), l.Txmgr.From())
dial.WaitRollupSync(l.shutdownCtx, l.Log, rollupClient, l1TargetBlock, time.Second*12)
```
It seems to me `CheckRecentTxs` should return the first block that contains the current nonce, indicating the block where the most recent tx was sent. Then we should sync to that target. Implementation [here](https://github.com/ethereum-optimism/optimism/pull/10193/files#diff-f9123497e13c8c45a184575da571357f48b0becd5e9aca7e0de54f34a0aa29f8R96)

File: op-batcher/batcher/driver.go (hunk 12 above)
Comment (bitwiseguy) [reply op-batcher/batcher/driver.go]: Moving function into `op-service/eth`

File: op-node/rollup/types.go (hunk 10 above)
Comment (bitwiseguy) [reply op-node/rollup/types.go]: Updated in https://github.com/ethereum-optimism/optimism/pull/10193/commits/ccaed7bca41f6a3112d5f28b4d6decf10a3bcfa9 to remove dependency on `op-node` and instead use an int `CheckRecentTxsDepth` flag as @sebastianst suggested.

Comment (bitwiseguy) [reply op-node/rollup/types.go]: @sebastianst Should we also add protection around the `CheckRecentTxsDepth` value so that it can't be set over a maximum limit of ~120? Since archival node support is needed to look back >128 blocks, maybe 120 is a nice safe value that should be supported by any normal node

File: op-batcher/batcher/driver.go [hunk 18]
```diff
@@ -324,6 +334,33 @@ func (l *BatchSubmitter) loop() {
 	}
 }
 
+// waitNodeSync Check to see if there was a batcher tx sent recently that
+// still needs more block confirmations before being considered finalized
+func (l *BatchSubmitter) waitNodeSync() error {
```
Comment (sebastianst) [inline op-batcher/batcher/driver.go:339]: Would be good to add a timeout to all but the last `WaitRollupSync` call, like
```
	ctx, cancel := context.WithTimeout(l.shutdownCtx, l.Config.NetworkTimeout)
	defer cancel()
```
and use it instead of `l.shutdownCtx`. We also do this for other short-lived network calls.

File: op-batcher/batcher/driver.go [hunk 19]
```diff
@@ -324,6 +334,33 @@ func (l *BatchSubmitter) loop() {
...
+		recentBlock, found, err := eth.CheckRecentTxs(l.shutdownCtx, l.Log, l.L1Client, l.Config.CheckRecentTxsDepth, l.Txmgr.From())
+		if err != nil {
+			return fmt.Errorf("failed when checking recent batcher txs: %w", err)
+		}
+		if found {
+			l1TargetBlock = recentBlock
+		}
+	}
+
+	return dial.WaitRollupSync(l.shutdownCtx, l.Log, rollupClient, l1TargetBlock, time.Second*12)
```
Comment (sebastianst) [inline op-batcher/batcher/driver.go:365]: Here it's probably fine to just use the shutdown ctx, because we can't estimate how long this takes.

File: op-node/service.go [hunk 20]
```diff
@@ -190,7 +190,6 @@ func NewConfigPersistence(ctx *cli.Context) node.ConfigPersistence {
 
 func NewDriverConfig(ctx *cli.Context) *driver.Config {
 	return &driver.Config{
-		VerifierConfDepth:   ctx.Uint64(flags.VerifierL1Confs.Name),
```
Comment (sebastianst) [inline op-node/service.go]: I believe this shouldn't be deleted.

File: op-node/service.go [hunk 21]
```diff
@@ -204,15 +203,15 @@ func NewRollupConfigFromCLI(log log.Logger, ctx *cli.Context) (*rollup.Config, e
...
+	rollupConfig, err := newRollupConfig(log, network, rollupConfigPath)
 	if err != nil {
 		return nil, err
 	}
 	applyOverrides(ctx, rollupConfig)
 	return rollupConfig, nil
 }
 
-func NewRollupConfig(log log.Logger, network string, rollupConfigPath string) (*rollup.Config, error) {
+func newRollupConfig(log log.Logger, network string, rollupConfigPath string) (*rollup.Config, error) {
```
Comment (sebastianst) [inline op-node/service.go]: since you reverted the rollup config change, I'd remove this diff.

File: op-service/eth/transactions.go [hunk 22]
```diff
@@ -42,3 +50,48 @@ func TransactionsToHashes(elems []*types.Transaction) []common.Hash {
...
+
+	oldestBlock := new(big.Int)
+	oldestBlock.Sub(currentBlock, big.NewInt(int64(depth)))
+	previousNonce, err := l1.NonceAt(ctx, addr, oldestBlock)
+	if err != nil {
+		return 0, false, fmt.Errorf("failed to retrieve previous nonce: %w", err)
+	}
+
+	if currentNonce == previousNonce {
+		lgr.Info("No recent batcher txs detected. No need to wait for rollup sync")
```
Comment (sebastianst) [inline op-service/eth/transactions.go]:
```suggestion
```
It's not in the batcher any more, so need to remove the batcher reference. But I actually suggest we just delete this log line, it's the last one remaining. Then we can also remove `lgr` from the function's parameters.

File: op-service/eth/transactions.go [hunk 23]
```diff
@@ -42,3 +50,48 @@ func TransactionsToHashes(elems []*types.Transaction) []common.Hash {
...
+		return oldestBlock.Uint64(), false, nil
+	}
+
+	lgr.Info("Recent tx detected. Searching for block with most recent tx")
+	// Decrease block num until we find the block before the most recent batcher tx was sent
+	targetNonce := currentNonce - 1
+	for currentNonce > targetNonce {
```
Comment (sebastianst) [inline op-service/eth/transactions.go]: I just realize, there's a very unlikely yet possible edge-case that a reorg happens while querying `previousNonce, err := l1.NonceAt(ctx, addr, oldestBlock)` where one tx would be added, so that the `currentNonce == previousNonce` check would fail, then another reorg removes that tx again, and then this loop would potentially run for a very long time. It's a very constructed case, but I think just to be on the safe side we should also test the current block number
```suggestion
	for currentNonce > targetNonce && currentBlock.Cmp(oldestBlock) != -1 {
```
We could add a test case for this that would currently fail where the first nonce query for block N returns x, then querying the nonce for N-depth returns x+1, then all subsequent queries return x again.
Maybe there are other edge cases too that reorgs might cause that I don't currently see.

**Review by sebastianst — CHANGES_REQUESTED** [review]
Looks pretty good now! Just a few minor things and one possible reorg edge-case that we might have missed so far.

File: op-batcher/batcher/driver.go (hunk 16 above)
Comment (sebastianst) [reply op-batcher/batcher/driver.go]: I think that's a good approach for now!

File: op-node/rollup/types.go (hunk 10 above)
Comment (sebastianst) [reply op-node/rollup/types.go]: Yes good idea, let's add a check to the batcher CLI config that this is at most 128. It doesn't make sense to use this functionality with higher values than 128, and realistic values are around 4.

File: op-node/rollup/types.go [hunk 24]
```diff
@@ -55,6 +55,7 @@ type Config struct {
 	Genesis Genesis `json:"genesis"`
 	// Seconds per L2 block
 	BlockTime uint64 `json:"block_time"`
+
```
Comment (sebastianst) [inline op-node/rollup/types.go]: it doesn't hurt, but changing this file is unrelated to this PR, maybe remove this diff.

File: op-batcher/flags/flags.go [hunk 25]
```diff
@@ -127,6 +127,20 @@ var (
 		Value:   2 * time.Minute,
 		EnvVars: prefixEnvVars("ACTIVE_SEQUENCER_CHECK_DURATION"),
 	}
+	CheckRecentTxsDepthFlag = &cli.IntFlag{
+		Name: "check-recent-txs-depth",
+		Usage: "Indicates how many blocks back the batcher should look during startup for a recent batch tx " +
+			"on L1. This should help avoid duplicate batcher txs.",
```
Comment (sebastianst) [inline op-batcher/flags/flags.go]:
```suggestion
			"on L1. This can speed up waiting for node sync. It should be set to the verifier confirmation depth of the sequencer (e.g. 4).",
```
the node sync avoids the duplicate batcher txs, while the recent tx check speeds up that sync waiting.

File: op-service/eth/transactions.go (hunk 23 above)
Comment (bitwiseguy) [reply op-service/eth/transactions.go]: I'm trying to understand the reorg situation you described. This is the nonce setting at each block before the reorg:
```
Blocks       495 496 497 498 499 500
Nonce          5   5   5   5   6   6
call NonceAt   x   -   -   -   -   x
```
Reorg occurs, removing tx that caused nonce to increment to 6. The nonce now looks like this:
```
Blocks       495 496 497 498 499 500
Nonce          5   5   5   5   5   5
call NonceAt   -   -   -   -   x   -
```
With the current implementation, we would start walking back from 500 to 499, then detect that the nonce at 499 is less than the nonce we previously read at 500, so `CheckForRecentTxs` would assume the nonce change occurred at block 500 and incorrectly return that value. It doesn't seem like the `currentBlock.Cmp(oldestBlock)` would help this situation. Did I describe the situation you had in mind correctly or am I missing something?

Comment (sebastianst) [reply op-service/eth/transactions.go]: I'm talking about two reorgs `depth` deep, so
* 500 -> 5
* now reorg a tx into block 495
* 495 -> 6
* now reorg this tx out again
* 499 -> 5
* 498 -> 5
* ... potentially going on with 5s for a long time.

File: op-batcher/flags/flags.go [hunk 26]
```diff
@@ -138,6 +152,7 @@ var requiredFlags = []cli.Flag{
 }
 
 var optionalFlags = []cli.Flag{
+	CheckRecentTxsDepthFlag,
```
Comment (sebastianst) [inline op-batcher/flags/flags.go:156]: forgot
```suggestion
	WaitNodeSyncFlag
	CheckRecentTxsDepthFlag,
```

File: op-service/eth/transactions.go (hunk 23 above)
Comment (bitwiseguy) [reply op-service/eth/transactions.go]: Updated in https://github.com/ethereum-optimism/optimism/pull/10193/commits/4abebbcc3e19a40c864a506fcc159e197e1d06ee

**Review by sebastianst — CHANGES_REQUESTED** [review]

File: op-batcher/flags/flags.go (hunk 26 above)
Comment (bitwiseguy) [reply op-batcher/flags/flags.go:156]: Fixed in https://github.com/ethereum-optimism/optimism/pull/10193/commits/173c91a18a50c3abeaf8fb6cfa77000dfcc0d644

**Review by sebastianst — APPROVED** [review]

File: op-node/rollup/types.go (hunk 10 above)
Comment (bitwiseguy) [reply op-node/rollup/types.go]: Fixed
